rust/agents/relayer/src/msg/metadata/aggregation.rs

```diff
@@ -0,0 +1,238 @@
+use async_trait::async_trait;
+use futures_util::future::{join_all, try_join};
+use std::ops::Deref;
+
+use derive_new::new;
+use eyre::Context;
+use tracing::{info, instrument};
+
+use hyperlane_core::{HyperlaneMessage, InterchainSecurityModule, H256, U256};
+
+use super::{BaseMetadataBuilder, MetadataBuilder};
+
+/// Bytes used to store one member of the (start, end) range tuple
+/// Copied from `AggregationIsmMetadata.sol`
+const METADATA_RANGE_SIZE: usize = 4;
+
+#[derive(Clone, Debug, new)]
+pub struct AggregationIsmMetadataBuilder {
+    base: BaseMetadataBuilder,
+}
+
+#[derive(Clone, Debug, new, PartialEq, Eq)]
+struct SubIsmMetadata {
+    /// The index of the sub-ISM in the aggregation ISM.
+    index: usize,
+    /// The metadata for the sub-ISM.
+    metadata: Vec<u8>,
+}
+
+impl AggregationIsmMetadataBuilder {
+    fn format_metadata(metadatas: &mut [SubIsmMetadata], ism_count: usize) -> Vec<u8> {
+        // See test solidity implementation of this fn at:
+        // https://github.com/hyperlane-xyz/hyperlane-monorepo/blob/445da4fb0d8140a08c4b314e3051b7a934b0f968/solidity/test/isms/AggregationIsm.t.sol#L35
+        fn encode_byte_index(i: usize) -> [u8; 4] {
+            (i as u32).to_be_bytes()
+        }
+        let range_tuples_size = METADATA_RANGE_SIZE * 2 * ism_count;
+        //  Format of metadata:
+        //  [????:????] Metadata start/end uint32 ranges, packed as uint64
+        //  [????:????] ISM metadata, packed encoding
+        // Initialize the range tuple part of the buffer, so the actual metadatas can
+        // simply be appended to it
+        let mut buffer = vec![0; range_tuples_size];
+        for SubIsmMetadata { index, metadata } in metadatas.iter_mut() {
+            let range_start = buffer.len();
+            buffer.append(metadata);
+            let range_end = buffer.len();
+
+            // The new tuple starts at the end of the previous ones.
+            // Also see: https://github.com/hyperlane-xyz/hyperlane-monorepo/blob/445da4fb0d8140a08c4b314e3051b7a934b0f968/solidity/contracts/libs/isms/AggregationIsmMetadata.sol#L49
+            let encoded_range_start = METADATA_RANGE_SIZE * 2 * (*index);
+            // Overwrite the 0-initialized buffer
+            buffer.splice(
+                encoded_range_start..(encoded_range_start + METADATA_RANGE_SIZE * 2),
+                [encode_byte_index(range_start), encode_byte_index(range_end)].concat(),
+            );
+        }
+        buffer
+    }
+
+    fn n_cheapest_metas(
+        mut metas_and_gas: Vec<(SubIsmMetadata, U256)>,
+        n: usize,
+    ) -> Vec<SubIsmMetadata> {
+        metas_and_gas.sort_by(|(_, gas_1), (_, gas_2)| gas_1.cmp(gas_2));
+        let mut cheapest: Vec<_> = metas_and_gas[..n].into();
+        cheapest.sort_by(|(meta_1, _), (meta_2, _)| meta_1.index.cmp(&meta_2.index));
+        cheapest.iter().cloned().map(|(meta, _)| meta).collect()
```

Comment:
can we add some comments here? wasn't obvious what was happening but now I realize we want to sort by gas, take first (threshold), and then make index monotonically increasing